### ISSUE TITLE
Possible issue with CommandQueue::downgrade

### ISSUE THREAD

**Issue by andrewhickman** [gfx-rs/gfx#2392]
It is possible to assign a transfer queue to a variable of graphics queue type.
```
let graphics: CommandQueue<B, Graphics> = unimplemented!();
let transfer: CommandQueue<B, Transfer> = unimplemented!();
*graphics.downgrade() = transfer;
```
Not sure if this is safety issue?

**Comment by omni-viral** [comment]
`Pin` would be the fix

**Comment by andrewhickman** [comment]
As I understand it, to use pinning here we would need to make `CommandQueue: !Unpin` and return `PinMut`. But then users would have to go through the unsafe `PinMut::get_mut` to do anything useful, at which point we may as well just make `CommandQueue::downgrade` unsafe!

**Comment by andrewhickman** [comment]
Vaguely relevant post https://www.ralfj.de/blog/2018/04/10/safe-intrusive-collections-with-pinning.html. 

After reading that I realized could use Pin at the cost of making all functions take CommandQueue by `PinMut`.

**Comment by omni-viral** [comment]
We shouldn't allow this downgrade because it is can trigger UB. The solution is to make it possible for wrapper to contain mutable reference to raw queue. And reconstruct wrapper during downgrading or even upgrading (that require checking of actual capabilities)

**Comment by kvark** [comment]
Fixed in #2948 :D

**Comment by omni-viral** [comment]
Если у вас нету тёти...
